cmd/quickdata/main.go

```diff
@@ -0,0 +1,468 @@
+	}
+	outPrintln()
+	// Bids (green), from best (highest) up to levels
+	bidCount := intMin(levels, len(b.Bids))
+	cumulativeBids := 0.0
+	for i := range bidCount {
+		lvl := b.Bids[i]
+		cumulativeBids += lvl.Amount
+		outPrintf("%s% -14.8f % -14.8f % -14.8f%s\n", ansiGreen, lvl.Price, lvl.Amount, cumulativeBids, ansiReset)
+	}
+}
+
+func renderTicker(t *ticker.Price) {
+	if t == nil {
+		return
+	}
+	spread := t.Ask - t.Bid
+	var spreadPct float64
+	if t.Last != 0 {
+		spreadPct = (spread / t.Last) * 100
+	}
+	outPrintf("%sLast:%s %.8f  %sBid:%s %.8f  %sAsk:%s %.8f\n",
+		ansiBold, ansiReset, t.Last, ansiGreen, ansiReset, t.Bid, ansiRed, ansiReset, t.Ask)
+	outPrintf("%sVol:%s  %.4f  %sSpread:%s %.8f (%.4f%%)  %sMark:%s %.8f  %sIndex:%s %.8f\n",
+		ansiDim, ansiReset, t.Volume, ansiDim, ansiReset, spread, spreadPct, ansiDim, ansiReset, t.MarkPrice, ansiDim, ansiReset, t.IndexPrice)
+}
+
+func renderTrades(trs []trade.Data) {
+	if len(trs) == 0 {
+		outPrintln("No trades.")
+		return
+	}
+	// Summaries
+	n := len(trs)
+	start := trs[0].Timestamp
+	end := trs[n-1].Timestamp
+	if end.Before(start) {
+		start, end = end, start
+	}
+	var baseVol, quoteVol float64
+	var buys, sells int
+	for i := range trs {
+		baseVol += trs[i].Amount
+		quoteVol += trs[i].Amount * trs[i].Price
+		if trs[i].Side&order.Buy == order.Buy || trs[i].Side&order.Bid == order.Bid || trs[i].Side&order.Long == order.Long {
+			buys++
+		} else if trs[i].Side&order.Sell == order.Sell || trs[i].Side&order.Ask == order.Ask || trs[i].Side&order.Short == order.Short {
+			sells++
+		}
+	}
+	vwap := 0.0
+	if baseVol != 0 {
+		vwap = quoteVol / baseVol
+	}
+	last := trs[n-1]
+	span := end.Sub(start)
+	outPrintf("%sTrades:%s N=%d Span=%s\n", ansiBold, ansiReset, n, span.Truncate(time.Second))
+	outPrintf("Range: %s -> %s\n", start.UTC().Format(time.RFC3339), end.UTC().Format(time.RFC3339))
+	outPrintf("VWAP: %.8f  Last: %.8f @ %s\n", vwap, last.Price, last.Timestamp.UTC().Format(time.RFC3339))
+	outPrintf("Volume: base=%.8f quote=%.8f  Buys/Sells: %d/%d\n", baseVol, quoteVol, buys, sells)
+}
+
+func renderKlines(kl []websocket.KlineData) {
+	if len(kl) == 0 {
+		outPrintln("No klines.")
+		return
+	}
+	n := len(kl)
+	start := kl[0].Timestamp
+	end := kl[n-1].Timestamp
+	if end.Before(start) {
+		start, end = end, start
+	}
+	firstOpen := kl[0].OpenPrice
+	lastClose := kl[n-1].ClosePrice
+	change := lastClose - firstOpen
+	changePct := 0.0
+	if firstOpen != 0 {
+		changePct = (change / firstOpen) * 100
+	}
+	high := kl[0].HighPrice
+	low := kl[0].LowPrice
+	var totalVol float64
+	for i := range kl {
+		if kl[i].HighPrice > high {
+			high = kl[i].HighPrice
+		}
+		if kl[i].LowPrice < low {
+			low = kl[i].LowPrice
+		}
+		totalVol += kl[i].Volume
+	}
+	avgVol := totalVol / float64(n)
+	span := end.Sub(start)
+	interval := kl[n-1].Interval
+	if interval == "" && n > 1 {
+		interval = kl[0].Interval
```

Comment:
Why could the first element will have the `kline interval` information but not the last `KlineData`? What am I missing here?